### PR TITLE
 [WIP/RFC] Only use Jedi, without worker/client/server/cache

### PR DESCRIPTION
**Pull request by blueyed**
I have tried to implement restarting in case
`g:deoplete#sources#jedi#python_path` gets changed, and figured that it
is way easier when using only Jedi.

There are likely previous features missing, and it needs cleaning up /
deleting still.

It uses the Jedi branch to improve handling of Environments:
davidhalter/jedi#1108

TODO:
- [ ] jedi.settings.additional_dynamic_modules
- [ ] revisit existing features, e.g. the boilerplate handling
- [ ] consider keeping the cache itself, only remove Worker/Client/Server
- [ ] allow to control environment using some variable, possibly wrapped into a command to select it (ref: https://github.com/davidhalter/jedi-vim/pull/836)

Works quite well already, tested with:

```python
import numpy

numpy.
```

```
2018-07-08 23:25:51,485 DEBUG    [14486] (deoplete.jedi) Line: 3, Col: 6, Filename: '…/vim/plugged/jedi-vim/pythonx/jedi/t.py'
2018-07-08 23:25:52,442 INFO     [14486] (deoplete.jedi) get_completions t = 956.456334ms, µ = 0, σ = 0)
2018-07-08 23:25:53,194 INFO     [14486] (deoplete.jedi) massage_completions t = 752.618779ms, µ = 0, σ = 0)
2018-07-08 23:25:53,195 INFO     [14486] (deoplete.jedi) gather_candidates t = 1712.006350ms, µ = 0, σ = 0)
2018-07-08 23:26:01,394 DEBUG    [14485] (deoplete.core) on_event: InsertEnter
2018-07-08 23:26:03,061 DEBUG    [14485] (deoplete.core) completion_begin: 'numpy.' (TextChangedI)
2018-07-08 23:26:12,425 DEBUG    [14485] (deoplete.core) on_event: InsertEnter
2018-07-08 23:26:12,944 DEBUG    [14485] (deoplete.core) completion_begin: '' (TextChangedI)
2018-07-08 23:26:13,003 DEBUG    [14485] (deoplete.core) completion_begin: 'n' (Async)
2018-07-08 23:26:14,475 DEBUG    [14485] (deoplete.core) completion_begin: 'numpy.' (TextChangedI)
2018-07-08 23:26:14,483 DEBUG    [14486] (deoplete.jedi) Line: 5, Col: 6, Filename: '…/vim/plugged/jedi-vim/pythonx/jedi/t.py'
2018-07-08 23:26:14,534 DEBUG    [14485] (deoplete.core) completion_begin: 'numpy.' (Async)
2018-07-08 23:26:14,702 INFO     [14486] (deoplete.jedi) get_completions t = 219.013144ms, µ = 0, σ = 0)
2018-07-08 23:26:15,270 INFO     [14486] (deoplete.jedi) massage_completions t = 567.887416ms, µ = 0, σ = 0)
2018-07-08 23:26:15,270 INFO     [14486] (deoplete.jedi) gather_candidates t = 791.059430ms, µ = 0, σ = 0)
```

So `massage_completions` takes a lot of the time by itself.

It might make certainly sense to add some caching on top, but it shouldn't be so complicated anymore.

/cc @tweekmonster

### CODE REVIEW

File: rplugin/python3/deoplete/sources/deoplete_jedi.py [hunk 1]
```diff
@@ -40,51 +48,82 @@ def on_init(self, context):
...
-        # aggressive cache.
-        # Two workers may be needed if working with very large source files.
-        self.worker_threads = vars.get(
-            'deoplete#sources#jedi#worker_threads', 2)
-        # Hard coded python interpreter location
-        self.python_path = vars.get(
-            'deoplete#sources#jedi#python_path', '')
+        # TODO
         self.extra_path = vars.get(
             'deoplete#sources#jedi#extra_path', [])
```
Comment (blueyed) [inline rplugin/python3/deoplete/sources/deoplete_jedi.py]: Jedi has `jedi.settings.additional_dynamic_modules` (used in jedi-vim).

**Comment by Shougo** [comment]
I think it is better version.
Because, deoplete is already parallel.

**Comment by blueyed** [comment]
Yep, I like it also so far.

File: rplugin/python3/deoplete/sources/deoplete_jedi.py [hunk 2]
```diff
@@ -168,155 +214,77 @@ def finalize_cached(self, cache_key, filters, cached):
...
+                return self.completion_dict(name, builtin_type, comp)
+
+        if type_ == 'class' and desc.startswith('builtins.'):
+            return self.completion_dict(name, type_, comp)
+
+        if type_ == 'function':
+            if comp.module_path not in cache and comp.line and comp.line > 1 \
+                    and os.path.exists(comp.module_path):
+                with open(comp.module_path, 'r') as fp:
+                    cache[comp.module_path] = fp.readlines()
```
[the remark this answers is not in the document]
Comment (blueyed) [reply rplugin/python3/deoplete/sources/deoplete_jedi.py]: Yes, likely only needed for some old and hopefully fixed misbehavior with regard to `@property` decorators.

**Comment by blueyed** [comment]
@Shougo 
You can pick it up, and either rebase it only locally first, or also push here.

But "Allow edits from maintainers." might not be enough.. :/

Ping @zchee.

Rebased it now, updated the submodules, but no other changes.

**Comment by blueyed** [comment]
@Shougo 
Maybe you want to create a new PR then - I won't have time for it currently.

**Comment by Shougo** [comment]
> Maybe you want to create a new PR then - I won't have time for it currently.

OK.  I will do it.

**Comment by Shougo** [comment]
I have created #192.

**Comment by blueyed** [comment]
Closing in favor of #192.

**Pull request closed without merging**